beacon_node/eth1/src/deposit_cache.rs

```diff
@@ -114,33 +175,111 @@ impl DepositCache {
     pub fn new(deposit_contract_deploy_block: u64) -> Self {
         DepositCache {
             deposit_contract_deploy_block,
+            finalized_block_height: deposit_contract_deploy_block.saturating_sub(1),
             ..Self::default()
         }
     }
 
-    /// Returns the number of deposits available in the cache.
+    pub fn from_deposit_snapshot(
+        deposit_contract_deploy_block: u64,
+        snapshot: &DepositTreeSnapshot,
+    ) -> Result<Self, String> {
+        let deposit_tree = DepositDataTree::from_snapshot(snapshot, DEPOSIT_TREE_DEPTH)
+            .map_err(|e| format!("Invalid DepositSnapshot: {:?}", e))?;
+        Ok(DepositCache {
+            logs: Vec::new(),
+            leaves: Vec::new(),
+            deposit_contract_deploy_block,
+            finalized_deposit_count: snapshot.deposit_count,
+            finalized_block_height: snapshot.execution_block_height,
+            deposit_tree,
+            deposit_roots: Vec::new(),
```

Comment:
I think this empty vec causes issues with the sanity check on the `deposit_roots` at startup. I got this CRIT after restarting my checkpoint synced node on Goerli/Prater:

> Sep 28 07:22:30.682 CRIT Failed to start beacon node             reason: Invalid SszDepositCache: deposit_roots length must be only one more than leaves. Leaves: 82, Deposit roots: 82

I added the lengths to see what was going on.

Usually we maintain an invariant that the 0th entry of `deposit_roots` is the root of the empty deposit tree, but I guess we may not want to do that anymore? I haven't thought enough about it.

Reply:
Ah very nice you found a bug when recovering from the snapshot!
> Usually we maintain an invariant that the 0th entry of deposit_roots is the root of the empty deposit tree

yes that invariant is still maintained everywhere except when recovering the cache from a snapshot (which was a bug). 
I've changed this to initialize the array with `vec![snapshot.deposit_root]`, which matches how the cache looks after finalization :)

I've also added a test to verify that cache recovery works with these conditions

Reply:
looks good!